Objects/tupleobject.c

```diff
@@ -333,38 +333,39 @@ tuplerepr(PyTupleObject *v)
     return NULL;
 }
 
-/* The addend 82520, was selected from the range(0, 1000000) for
-   generating the greatest number of prime multipliers for tuples
-   up to length eight:
 
-     1082527, 1165049, 1082531, 1165057, 1247581, 1330103, 1082533,
-     1330111, 1412633, 1165069, 1247599, 1495177, 1577699
-
-   Tests have shown that it's not worth to cache the hash value, see
+/* Tests have shown that it's not worth to cache the hash value, see
    issue #9685.
-*/
 
+   The structure is that of a standard Bernstein "DJBX33A" hash.
+   We compute the hash with unsigned variables for well-defined
+   overflow behavior. */
 static Py_hash_t
 tuplehash(PyTupleObject *v)
 {
-    Py_uhash_t x;  /* Unsigned for defined overflow behavior. */
-    Py_hash_t y;
-    Py_ssize_t len = Py_SIZE(v);
-    PyObject **p;
-    Py_uhash_t mult = _PyHASH_MULTIPLIER;
-    x = 0x345678UL;
-    p = v->ob_item;
-    while (--len >= 0) {
-        y = PyObject_Hash(*p++);
-        if (y == -1)
+    /* Arbitrary starting value, any non-zero value would work. The
+       value here was chosen to retain the historical hash of the empty
+       tuple, which equals the initial value of this algorithm. */
+    Py_uhash_t x = 3527539;
+    Py_uhash_t MUL = _PyHASH_MULTIPLIER2, ERR = -1;
+    Py_ssize_t i, len = Py_SIZE(v);
+    PyObject **item = v->ob_item;
+    for (i = 0; i < len; i++) {
+        Py_uhash_t t = PyObject_Hash(item[i]);
+        if (t == ERR) {
             return -1;
-        x = (x ^ y) * mult;
-        /* the cast might truncate len; that doesn't change hash stability */
-        mult += (Py_hash_t)(82520UL + len + len);
```

Comment:
Why did the length adjustment get dropped? This helps may tuple hashes distinct when they differ only by the presence of empty elements.